### ISSUE TITLE
Intellisense for Symbol properties

### ISSUE THREAD

**Issue by joelday** [microsoft/TypeScript#54870]
# Suggestion

## 🔍 Search Terms

With and without "is:issue is:open"
"symbol intellisense"
"symbol autocomplete"

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Intellisense provides a list of symbol properties which, when accepted, insert indexed accessor syntax for that symbol. This list would be comprised of known symbols within the current scope or those that can be with an import statement.

## 📃 Motivating Example

This example demonstrates a type that implements the global `Disposable` interface introduced by TC39 ECMAScript Explicit Resource Management. (PR#54505)

```typescript
class ExampleClass implements Disposable {
  [Symbol.dispose](): void {
    // ...
  }
}

const example = new ExampleClass();

example[Symbol.dispose]();
```

## 💻 Use Cases

TL;DR: `Symbol.dispose` and `Symbol.asyncDispose` are the first well-known symbols that are useful to call directly, and I want to easily know that they're there.

Typically, ES6 Symbols are intended as an implementation detail, but the new global `Disposable` interface, however, is analogous to this very prolific interface:
```typescript
interface Disposable {
  dispose(): void;
}
```

In order to support `using` syntax, I need to implement the global `Disposable` interface. I have a few options:

- Switch to the global `Disposable` interface, but lose the visibility to developers of the lifecycle management requirements of implementations.
- Make my `Disposable` interface extend the global `Disposable` interface, or possibly merge it with the global one and then, by convention, do the following:
```typescript
dispose(): void {
  // ...
}

[Symbol.dispose](): void {
  this.dispose();
}
```
An option is to create a utility class decorator that does that for me, but unfortunately that creates the issue of how to make that class recognized as an implementation of the global `Disposable`. (#4881)

**Comment by Andarist** [comment]
Related to https://github.com/microsoft/TypeScript/issues/28364 and https://github.com/microsoft/TypeScript/issues/52268#issuecomment-1556007419

**Comment by joelday** [comment]
@Andarist happy to make a PR for these, whatever the desired behavior is.

**Comment by Andarist** [comment]
I have no real say on those matters really. I think though that:
- it would be good to have *concrete* suggestions for `Symbol.iterator`, `Symbol.disposable` etc instead of just `Symbol`
- those should have a low priority and should appear at the bottom of the list
- cursor placement after selecting the completion should make sense. Since the proposal is to offer auto-completions for concrete symbols then perhaps no changes would be needed here as the cursor already is positioned *after* the closing bracket

**Comment by Erbenos** [comment]
Another good application for this is declaration merging of interfaces of libraries, where eg.: new properties added by library consumers would be symbols, allowing for some level of future-proofing against clashes with newer iterations of library API.
